README
NAME
    Module::Package::RDF - drive your distribution with RDF

SYNOPSIS
    In your Makefile.PL:

      use inc::Module::Package 'RDF';

    That's all folks!

DESCRIPTION
    Really simple Makefile.PL.

FLAVOURS
    Currently this module only defines the ":standard" flavour.

  :standard
    This is the default, so the following are equivalent:

      use inc::Module::Package 'RDF';
      use inc::Module::Package 'RDF:standard';

    In addition to the inherited behavior, this flavour uses the following
    plugins:

    *   AutoLicense

    *   AutoManifest

    *   Copyright

    *   DOAP

    *   DOAPChangeSets

    *   RDF

    *   ReadmeFromPod

    *   TrustMetaYml

    And sets "static_config" and "sign".

BUGS
    Please report any bugs to <http://rt.cpan.org/>.

SEE ALSO
    Module::Package, Module::Install::RDF, Module::Install::DOAP,
    Module::Install::DOAPChangeSets .

    <http://www.perlrdf.org/>.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENSE
    Copyright (C) 2011-2013 by Toby Inkster

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
Module-Package-RDF
==================

Created:      2011-10-01
Home page:    <https://metacpan.org/release/Module-Package-RDF>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Module-Package-RDF>
Maintainer:   Toby Inkster <mailto:tobyink@cpan.org>

0.013  2013-03-19

 - (Update) Bump versions of several dependencies.
 - (Update) Switch to doap-deps vocab for expressing project dependencies
   (mkdist).

0.012  2013-03-19

 - (Addition) use Module::Install::Changes

0.011  2013-03-16

 - use Module::Install::RDF's new write_meta_ttl feature

0.010  2013-03-11

 - (Bugfix) Avoid auto-deps thingy from picking up
   Module::Install::StandardDocuments as a dep - it's bundled
   Fixes RT#82158
   ++$CHORNY
 - use Module::Install::Copyright

0.009  2012-09-25

 - (Addition) Module::Install::StandardDocuments.
 - App::mkdist now pulls author tests from ~/perl5/xt rather than producing
   hard-coded ones.
 - bin/mkdist now sports a --package_flavour option.

0.008  2012-07-05

 - (Bugfix) Get t/01basic.t working again.

0.007  2012-07-05  Bond, James Bond

 - (Packaging Update) Bump versions of some dependencies.

0.006  2012-07-04

 - (Update) mkdist now creates Pretdsl instead of Turtle.
 - Module::Package::RDF::Create is now App::mkdist.
 - The ':standard' bit of 'RDF:standard' is now optional.

0.005  2011-12-03

 - (Bugfix) use base
 - (Documentation) add 'DISCLAIMER OF WARRANTIES' section
 - (Documentation) add 'NAME' section

0.004  2011-11-21

 - (Bugfix) Module::Package::RDF::Create now handles '--use boolean' and
   '--use autodie' specially.
 - (Documentation) The 'mkdist' functionality should be split into a
   different release.
 - (Regression) use parent

0.003  2011-10-06

 - (Addition) Generated modules will include $AUTHORITY variables.
 - (Addition) Generated modules will include DISCLAIMER OF WARRANTIES in
   pod.
 - (Addition) Generated modules will include SYNOPSIS in pod.
 - (Addition) Module::Package::Dist::RDF::standard to run code at user end.
 - (Bugfix) Module::Package::Dist::RDF::standard uses
   Meta::Install::TrustMetaYml - this resolves some dependency issues with
   resultant packages.
 - (Bugfix) Prevent Module::Package::RDF::Create's pod and the template
   module's pod from getting muddled.
 - (Update) Generated modules now default to using perl 5.010.
 - (Update) use Module::Package 0.30

0.002  2011-10-03

 - (Addition) Include mkdist script and Module::Package::RDF::Create.

0.001  2011-10-01  Initial release




lib/Module/Package/RDF.pm
package Module::Package::RDF;

our $VERSION = '0.002';

package Module::Package::RDF::standard;

use 5.008003;
use strict;

use RDF::Trine 0.133 ();
use Module::Package 0.29 ();
use Module::Install::AutoLicense 0.08 ();
use Module::Install::ReadmeFromPod 0.12 ();
use Module::Install::RDF 0.001 ();
use Module::Install::DOAP 0.001 ();
use Module::Install::DOAPChangeSets 0.103 ();
use Log::Log4perl 0 qw(:easy);

our $VERSION = '0.002';

use Moo;
extends 'Module::Package::Plugin';

sub main
{
	my ($self) = @_;

	$self->mi->rdf_metadata;
	$self->mi->doap_metadata;
	$self->mi->sign;

	# These run later, as specified.
	$self->post_all_from(sub {Log::Log4perl->easy_init($ERROR);$self->mi->write_doap_changes});
	$self->post_all_from(sub {$self->mi->auto_license});
	$self->post_all_from(sub {$self->mi->auto_manifest});
	$self->post_all_from(sub {$self->mi->auto_install});
	
	$self->post_all_from(sub {$self->mi->clean_files('Changes')});
	$self->post_all_from(sub {$self->mi->clean_files('MANIFEST')});
	$self->post_all_from(sub {$self->mi->clean_files('SIGNATURE')});
	$self->post_all_from(sub {$self->mi->clean_files('README')});
	$self->post_all_from(sub {$self->mi->clean_files('LICENSE')});
	$self->post_all_from(sub {$self->mi->clean_files('META.yml')});
}

# We don't want to auto-invoke all_from...
sub all_from
{
	my $self = shift;
	# $self->mi->_all_from(@_);
	$_->() for @{$self->{post_all_from} || []};
}

sub write_deps_list {}

=head1 SYNOPSIS

In your Makefile.PL:

  use inc::Module::Package 'RDF:standard';

That's all folks!

=head1 DESCRIPTION

Really simple Makefile.PL.

=head1 FLAVOURS

Currently this module only defines the C<:standard> flavour.

=head2 :standard

In addition to the inherited behavior, this flavour uses the following plugins:

=over

=item * RDF

=item * DOAP

=item * DOAPChangeSets

=item * ReadmeFromPod

=item * AutoLicense

=item * AutoManifest

=back

=head1 BUGS

Please report any bugs to L<http://rt.cpan.org/>.

=head1 SEE ALSO

L<Module::Package>,
L<Module::Install::RDF>,
L<Module::Install::DOAP>,
L<Module::Install::DOAPChangeSets> .

L<http://www.perlrdf.org/>.

=head1 AUTHOR

Toby Inkster E<lt>tobyink@cpan.orgE<gt>.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Toby Inkster

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut


lib/Module/Package/RDF/Create.pm
package Module::Package::RDF::Create;

use common::sense;
our $VERSION = 0.002;

use Carp;
use DateTime;
use IO::All;
use Software::License;
use Text::Template;
use URI::Escape qw[];

{
	my %templates;
	sub import
	{
		my ($class) = @_;
		
		my $key = undef;
		while (my $line = <DATA>)
		{
			if ($line =~ /^COMMENCE (.+)$/)
			{
				$key = $1;
			}
			else
			{
				$templates{$key} .= $line;
			}
		}
	}
	sub get_template
	{
		my ($class, $key) = @_;
		return Text::Template->new(-type=>'string', -source=>$templates{$key});
	}
	sub get_template_names
	{
		my ($class) = @_;
		return keys %templates;
	}
}

sub fill_in_template
{
	my ($self, $template) = @_;
	$template = $self->get_template($template) unless ref $template;
	
	my %hash;
	while (my ($k, $v) = each %$self)
	{
		$hash{$k} = ref $v ? \$v : $v;
	}
	
	return $template->fill_in(
		-hash => \%hash,
		);
}

sub iofile
{
	my ($self, $file) = @_;
	my $basedir = sprintf($self->{destination}, $self->{dist_name});
	return io($basedir.$file)->assert;
}

sub create
{
	my ($class, $name, %options) = @_;
	
	$options{module_name} = $name;
	$options{dist_name}   = $name;
	
	if ($name =~ /::/)
	{
		$options{dist_name} =~ s/::/-/g;
	}
	elsif ($name =~ /\-/)
	{
		$options{module_name} =~ s/\-/::/g;
	}
	
	my %use;
	if ($options{use})
	{
		%use = map {$_=>1} @{$options{use}};
	}
	$options{use} = \%use;
	
	my $self = bless \%options, $class;
	
	$self->set_defaults;
	$self->create_module;
	$self->create_makefile_pl;
	$self->create_metadata;
	$self->create_tests;
	$self->create_author_tests;
}

sub set_defaults
{
	my ($self) = @_;
	
	croak "Need an author name."   unless defined $self->{author}{name};
	croak "Need an author cpanid." unless defined $self->{author}{cpanid};

	$self->{author}{mbox} ||= sprintf('%s@cpan.org', $self->{author}{cpanid});

	$self->{backpan} ||= sprintf('http://backpan.cpan.org/authors/id/%s/%s/%s/',
		substr(uc $self->{author}{cpanid}, 0, 1),
		substr(uc $self->{author}{cpanid}, 0, 2),
		uc $self->{author}{cpanid},
		);
	
	$self->{abstract} ||= 'a module that does something-or-other';
	$self->{version}  ||= '0.001';
	
	$self->{version_ident} = 'v_'.$self->{version};
	$self->{version_ident} =~ s/\./-/g;
	
	$self->{destination} ||= './%s/';
	
	unless ($self->{module_filename})
	{
		$self->{module_filename} = 'lib::'.$self->{module_name};
		$self->{module_filename} =~ s/::/\//g;
		$self->{module_filename} .= '.pm';
	}
	
	$self->{copyright}{holder} ||= $self->{author}{name};
	$self->{copyright}{year}   ||= DateTime->now->year;
	
	$self->{licence_class} ||= 'Software::License::Perl_5'; 
	eval sprintf('use %s;', $self->{licence_class});
	$self->{licence} = $self->{licence_class}->new({
		year    => $self->{copyright}{year},
		holder  => $self->{copyright}{holder},
		});
	
	$self->{includes} = [grep {!/^(common_sense|strict|warnings|moose|5.008|5.010|namespace_clean)$/} keys %{$self->{use}}];
	
	{
		my @mr = @{ $self->{includes} };
		push @mr, 'Moose'            if $self->{use}{moose};
		push @mr, 'common::sense'    if $self->{use}{common_sense};
		push @mr, 'namespace::clean' if $self->{use}{namespace_clean}
		                             || $self->{use}{moose};
		
		if (@mr)
		{
			$self->{requires} = sprintf(";\n\t:requires %s",
				(join ' , ', (map { "\"$_\"" } @mr))
				);
		}
		else
		{
			$self->{requires} = '';
		}
	}
	
	$self->{pragmas} ||= join "\n", do {
			my @pragmas;
			push @pragmas, ($self->{use}{'5.010'} ? '5.010' : '5.008');
			push @pragmas, ($self->{use}{common_sense} ? ('common::sense') : ('strict','warnings'));
			push @pragmas, 'constant { FALSE => 0, TRUE => 1 }';
			push @pragmas, 'utf8';
			push @pragmas, 'Moose' if $self->{use}{moose};
			map { sprintf('use %s;', $_) } @pragmas;
		};
	
	$self->{final_pragmas} ||= join "\n", do {
			my @pragmas;
			push @pragmas, 'namespace::clean' if $self->{use}{namespace_clean} || $self->{use}{moose};
			map { sprintf('use %s;', $_) } @pragmas;
		};

	$self->{final_code} ||= join "\n", do {
			my @lines;
			push @lines, '__PACKAGE__->meta->make_immutable;' if $self->{use}{moose};
			push @lines, 'TRUE;';
			@lines;
		};
		
	foreach (qw(pragmas final_pragmas includes))
	{
		if (ref $self->{$_} eq 'ARRAY')
		{
			$self->{$_} = join "\n", map { "use $_;" } @{ $self->{$_} };
		}
	}
}

sub create_module
{
	my ($self) = @_;
	
	$self->iofile( $self->{module_filename} )
		->print( $self->fill_in_template('module') );
}

sub create_makefile_pl
{
	my ($self) = @_;
	
	$self->iofile('Makefile.PL')
		->print($self->fill_in_template('Makefile.PL'));
}

sub create_metadata
{
	my ($self) = @_;
	
	$self->iofile($_)
		->print($self->fill_in_template($_))
		foreach grep { m#^meta/# } $self->get_template_names;
}

sub create_tests
{
	my ($self) = @_;
	
	$self->iofile($_)
		->print($self->fill_in_template($_))
		foreach grep { m#^t/# } $self->get_template_names;
}

sub create_author_tests
{
	my ($self) = @_;
	
	$self->iofile($_)
		->print($self->fill_in_template($_))
		foreach grep { m#^xt/# } $self->get_template_names;
}

=head1 NAME

Module::Package::RDF::Create - create distributions that will use Module::Package::RDF.

=head1 SYNOPSIS

  mkdist Local::Example::Useful

=head1 DESCRIPTION

=over

=item C<< Module::Package::RDF::Create->create($distname, %options) >>

Create a distribution directory including all needed files.

=back

=head1 SEE ALSO

L<Module::Package::RDF>.

=head1 AUTHOR

Toby Inkster E<lt>tobyink@cpan.orgE<gt>.

=head1 COPYRIGHT AND LICENCE

This software is copyright (c) 2011 by Toby Inkster.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

1;

__DATA__

COMMENCE module
package {$module_name};

{$pragmas}

our $VERSION = '{$version}';

{$includes}

{$final_pragmas}

# Your code goes here

{$final_code}

=head1 NAME

{$module_name} - {$abstract}

=head1 DESCRIPTION

=head1 BUGS

Please report any bugs to
L<http://rt.cpan.org/Dist/Display.html?Queue={URI::Escape::uri_escape($dist_name)}>.

=head1 SEE ALSO

=head1 AUTHOR

{$author->{name}} E<lt>{$author->{mbox}}E<gt>.

=head1 COPYRIGHT AND LICENCE

{$licence->notice}

COMMENCE Makefile.PL
use inc::Module::Package 'RDF:standard';

COMMENCE meta/changes.ttl
# This file acts as the project's changelog.

@prefix :        <http://usefulinc.com/ns/doap#> .
@prefix dcs:     <http://ontologi.es/doap-changeset#> .
@prefix dc:      <http://purl.org/dc/terms/> .
@prefix dist:    <http://purl.org/NET/cpan-uri/dist/{URI::Escape::uri_escape($dist_name)}/> .
@prefix rdfs:    <http://www.w3.org/2000/01/rdf-schema#> .
@prefix xsd:     <http://www.w3.org/2001/XMLSchema#> .

dist:project :release dist:{$version_ident} .
dist:{$version_ident}
	a               :Version ;
	dc:issued       "{DateTime->now->ymd('-')}"^^xsd:date ;
	:revision       "{$version}"^^xsd:string ;
	:file-release   <{$backpan}{URI::Escape::uri_escape($dist_name)}-{$version}.tar.gz> ;
	rdfs:label      "Initial release" .

COMMENCE meta/doap.ttl
# This file contains general metadata about the project.

@prefix :        <http://usefulinc.com/ns/doap#> .
@prefix dc:      <http://purl.org/dc/terms/> .
@prefix foaf:    <http://xmlns.com/foaf/0.1/> .
@prefix rdfs:    <http://www.w3.org/2000/01/rdf-schema#> .
@prefix xsd:     <http://www.w3.org/2001/XMLSchema#> .

<http://purl.org/NET/cpan-uri/dist/{URI::Escape::uri_escape($dist_name)}/project>
	a               :Project ;
	:programming-language "Perl" ;
	:name           "{$dist_name}" ;
	:homepage       <https://metacpan.org/release/{URI::Escape::uri_escape($dist_name)}> ;
	:download-page  <https://metacpan.org/release/{URI::Escape::uri_escape($dist_name)}> ;
	:bug-database   <http://rt.cpan.org/Dist/Display.html?Queue={URI::Escape::uri_escape($dist_name)}> ;
	:created        "{DateTime->now->ymd('-')}"^^xsd:date ;
	:license        <{$licence->url}> ;
	:developer      [ a foaf:Person ; foaf:name "{$author->{name}}" ; foaf:mbox <mailto:{$author->{mbox}}> ] .

<{$licence->url}>
	dc:title        "{$licence->name}" .

COMMENCE meta/makefile.ttl
# This file provides instructions for packaging.

@prefix : <http://purl.org/NET/cpan-uri/terms#> .

<http://purl.org/NET/cpan-uri/dist/{URI::Escape::uri_escape($dist_name)}/project>
	:perl_version_from _:main ;
	:version_from _:main ;
	:readme_from _:main ;
	:test_requires "Test::More 0.61" {$requires} .

_:main <http://www.semanticdesktop.org/ontologies/2007/03/22/nfo#fileName> "{$module_filename}" .

COMMENCE t/01basic.t
use Test::More tests => 1;
BEGIN \{ use_ok('{$module_name}') \};

COMMENCE xt/01pod.t
use Test::More;
eval "use Test::Pod 1.00";
plan skip_all => "Test::Pod 1.00 required for testing POD" if $@;
all_pod_files_ok();

COMMENCE xt/02pod_coverage.t
use Test::More;
use Test::Pod::Coverage;

my @modules = qw({$module_name});
pod_coverage_ok($_, "$_ is covered")
	foreach @modules;
done_testing(scalar @modules);

COMMENCE xt/03meta_uptodate.t
use Test::More tests => 1;
use Test::RDF::DOAP::Version;
doap_version_ok('{$dist_name}', '{$module_name}');
